File: NotesApp/src/screens/NotesScreen.js
import React, { useState, useEffect, useContext } from "react";
import {
  View, Text, FlatList, StyleSheet, TouchableOpacity, ActivityIndicator,
} from "react-native";
import { getNotes, deleteNote, updateNote } from "../services/note.service";
import { AuthContext } from "../contexts/AuthContext";
import NoteItem from "../components/NoteItem";
import AddNoteModal from "../components/AddNoteModal";
import EditNoteModal from "../components/EditNoteModal";

const NotesScreen = () => {
  const [notes, setNotes] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const [addModalVisible, setAddModalVisible] = useState(false);
  const [editingNote, setEditingNote] = useState(null);
  const { user } = useContext(AuthContext);

  const fetchNotes = async () => {
    setLoading(true);
    try {
      const fetchedNotes = await getNotes(user?.$id);
      setNotes(fetchedNotes);
    } catch (err) {
      setError("Failed to load notes. Please try again.");
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    if (user) fetchNotes();
  }, [user]);

  const handleNoteAdded = (newNote) => {
    setNotes((current) => [newNote, ...current]);
  };

  const handleNoteDeleted = (noteId) => {
    setNotes((current) => current.filter((n) => n.$id !== noteId));
  };

  const handleNoteUpdated = (updatedNote) => {
    setNotes((current) => current.map((n) => (n.$id === updatedNote.$id ? updatedNote : n)));
  };

  const renderEmptyComponent = () => (
    <View style={styles.emptyContainer}>
      <Text style={styles.emptyText}>You don't have any notes yet.</Text>
      <Text style={styles.emptySubtext}>Tap the + button to create your first note!</Text>
    </View>
  );

  if (loading && notes.length === 0) {
    return <View style={styles.centered}><ActivityIndicator size="large" color="#007BFF" /></View>;
  }

  if (error && notes.length === 0) {
    return <View style={styles.centered}><Text style={styles.errorText}>{error}</Text></View>;
  }

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.title}>My Notes</Text>
        <TouchableOpacity style={styles.addButton} onPress={() => setAddModalVisible(true)}>
          <Text style={styles.addButtonText}>+ Add Note</Text>
        </TouchableOpacity>
      </View>

      <FlatList
        data={notes}
        keyExtractor={(item) => item.$id}
        renderItem={({ item }) => (
          <NoteItem
            note={item}
            onNoteDeleted={handleNoteDeleted}
            onEdit={() => setEditingNote(item)}
          />
        )}
        contentContainerStyle={notes.length === 0 ? { flex: 1 } : {}}
        ListEmptyComponent={!loading && renderEmptyComponent()}
        refreshing={loading}
        onRefresh={fetchNotes}
      />

      <AddNoteModal
        visible={addModalVisible}
        onClose={() => setAddModalVisible(false)}
        onNoteAdded={handleNoteAdded}
        userId={user?.$id}
      />

      <EditNoteModal
        visible={!!editingNote}
        onClose={() => setEditingNote(null)}
        onNoteUpdated={handleNoteUpdated}
        note={editingNote}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: "#f5f5f5" },
  header: { flexDirection: "row", justifyContent: "space-between", alignItems: "center", padding: 16 },
  title: { fontSize: 24, fontWeight: "bold", color: "#333" },
  addButton: { backgroundColor: "#2196F3", paddingVertical: 8, paddingHorizontal: 12, borderRadius: 5 },
  addButtonText: { color: "white", fontWeight: "bold" },
  centered: { flex: 1, justifyContent: "center", alignItems: "center" },
  errorText: { color: "red", fontSize: 16, textAlign: "center" },
  emptyContainer: { flex: 1, justifyContent: "center", alignItems: "center", padding: 20 },
  emptyText: { fontSize: 18, fontWeight: "bold", marginBottom: 10 },
  emptySubtext: { fontSize: 16, color: "#666", textAlign: "center" },
});

export default NotesScreen;

File: notes_app/lib/screens/notes_screen.dart
import 'package:flutter/material.dart';
import 'package:appwrite/models.dart';
import 'package:provider/provider.dart';
import '../services/note_service.dart';
import '../providers/auth_provider.dart';
import '../widgets/note_item.dart';
import '../widgets/add_note_modal.dart';
import '../widgets/edit_note_modal.dart';

class NotesScreen extends StatefulWidget {
  const NotesScreen({Key? key}) : super(key: key);

  @override
  _NotesScreenState createState() => _NotesScreenState();
}

class _NotesScreenState extends State<NotesScreen> {
  final NoteService _noteService = NoteService();
  List<Document> _notes = [];
  bool _isLoading = true;
  String? _error;

  @override
  void initState() {
    super.initState();
    _fetchNotes();
  }

  Future<void> _fetchNotes() async {
    final authProvider = Provider.of<AuthProvider>(context, listen: false);
    setState(() { _isLoading = true; _error = null; });
    try {
      final fetchedNotes = await _noteService.getNotes(userId: authProvider.user?.$id);
      setState(() { _notes = fetchedNotes; _isLoading = false; });
    } catch (e) {
      setState(() { _error = 'Failed to load notes. Please try again.'; _isLoading = false; });
    }
  }

  void _showAddNoteDialog() {
    final authProvider = Provider.of<AuthProvider>(context, listen: false);
    showDialog(
      context: context,
      builder: (context) => AddNoteModal(
        userId: authProvider.user?.$id ?? '',
        onNoteAdded: (newNote) => setState(() => _notes = [newNote, ..._notes]),
      ),
    );
  }

  void _showEditNoteDialog(Document note) {
    showDialog(
      context: context,
      builder: (context) => EditNoteModal(
        note: note,
        onNoteUpdated: (updatedNote) {
          setState(() {
            _notes = _notes.map((n) => n.$id == updatedNote.$id ? updatedNote : n).toList();
          });
        },
      ),
    );
  }

  void _handleNoteDeleted(String noteId) {
    setState(() { _notes = _notes.where((n) => n.$id != noteId).toList(); });
  }

  Widget _buildEmptyView() {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          const Icon(Icons.note_add, size: 64, color: Colors.grey),
          const SizedBox(height: 16),
          const Text("You don't have any notes yet.", style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
          const SizedBox(height: 10),
          Text("Tap the + button to create your first note!", style: TextStyle(fontSize: 16, color: Colors.grey[600]), textAlign: TextAlign.center),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('My Notes')),
      body: _isLoading && _notes.isEmpty
          ? const Center(child: CircularProgressIndicator())
          : _error != null && _notes.isEmpty
              ? Center(child: Text(_error!, style: const TextStyle(color: Colors.red)))
              : _notes.isEmpty
                  ? _buildEmptyView()
                  : RefreshIndicator(
                      onRefresh: _fetchNotes,
                      child: ListView.builder(
                        padding: const EdgeInsets.all(12),
                        itemCount: _notes.length,
                        itemBuilder: (context, index) {
                          return NoteItem(
                            note: _notes[index],
                            onNoteDeleted: _handleNoteDeleted,
                            onEdit: () => _showEditNoteDialog(_notes[index]),
                          );
                        },
                      ),
                    ),
      floatingActionButton: FloatingActionButton(
        onPressed: _showAddNoteDialog,
        child: const Icon(Icons.add),
      ),
    );
  }
}
